ethereum_bridge/src/test_utils.rs

```diff
@@ -0,0 +1,144 @@
+//! Test utilies for the Ethereum bridge crate.
+
+use std::collections::{BTreeSet, HashMap};
+
+use borsh::BorshSerialize;
+use namada_core::ledger::storage::testing::TestStorage;
+use namada_core::types::address::{self, Address};
+use namada_core::types::key::{
+    self, protocol_pk_key, RefTo, SecretKey, SigScheme,
+};
+use namada_core::types::token;
+use namada_proof_of_stake::epoched::Epoched;
+use namada_proof_of_stake::parameters::PosParams;
+use namada_proof_of_stake::types::{
+    ValidatorConsensusKeys, ValidatorEthKey, ValidatorSet, WeightedValidator,
+};
+use namada_proof_of_stake::PosBase;
+use rand::prelude::ThreadRng;
+use rand::thread_rng;
+
+/// Validator keys used for testing purposes.
+pub struct TestValidatorKeys {
+    /// Consensus keypair.
+    pub consensus: key::common::SecretKey,
+    /// Protocol keypair.
+    pub protocol: key::common::SecretKey,
+    /// Ethereum hot keypair.
+    pub eth_bridge: key::common::SecretKey,
+    /// Ethereum cold keypair.
+    pub eth_gov: key::common::SecretKey,
+}
+
+/// Set up a [`TestStorage`] initialized at genesis with a single
+/// validator.
+///
+/// The validator's address is [`address::testing::established_address_1`].
+#[inline]
+pub fn setup_default_storage()
+-> (TestStorage, HashMap<Address, TestValidatorKeys>) {
+    setup_storage_with_validators(HashMap::from_iter([(
+        address::testing::established_address_1(),
+        100_u64.into(),
+    )]))
+}
+
+/// Set up a [`TestStorage`] initialized at genesis with the given
+/// validators.
+pub fn setup_storage_with_validators(
+    active_validators: HashMap<Address, token::Amount>,
+) -> (TestStorage, HashMap<Address, TestValidatorKeys>) {
+    // set last height to a reasonable value;
+    // it should allow vote extensions to be cast
+    let mut storage = TestStorage {
+        last_height: 3.into(),
```

Comment:
nit: this `3` could be a const somewhere with the something similar to the comment above as a docstring - what's the minimum block height for which vote extensions could be cast during the consensus round? (2 or 3?)

Reply:
3 was chosen somewhat arbitrarily. you could have already sent a vote extension at `last_height = 1`, even

Reply:
ah right, I thought there might have been something specific about the `last_height` value being `3` (or rather, not being `1`)